application/dash_application/utility/plot_builder.py

```diff
@@ -769,4 +772,247 @@ def items_for_df(self):
             sample_type_col: self.sample_types,
             PINERY_COL.TissuePreparation: self.tissue_materials,
             COMMON_COL.Reference: self.reference,
-        }
\ No newline at end of file
+        }
+
+
+class Subplot:
+    def __init__(
+            self,
+            title,
+            df,
+            x_col,
+            y_col,
+            y_label,
+            colourby,
+            shapeby,
+            hovertext_cols,
+            cutoff_lines,
+            name_col,
+            markermode,
+            bar_positive,
+            bar_negative,
+            log_y
+    ):
+        self.title = title
+        self.y_label = y_label
+        self.df = df
+        self.x_col = x_col
+        self.y_col = y_col
+        self.colourby = colourby
+        self.shapeby = shapeby
+        self.hovertext_cols = hovertext_cols
+        self.cutoff_lines = [] if cutoff_lines is None else cutoff_lines
+        self.name_col = name_col
+        self.markermode = markermode
+        self.bar_positive = bar_positive
+        self.bar_negative = bar_negative
+        self.log_y = log_y
+
+    def traces(self):
+        return _generate_traces(
+            self.df,
+            self.x_col,
+            self.y_col,
+            self.colourby,
+            self.shapeby,
+            self.hovertext_cols,
+            self.cutoff_lines,
+            self.name_col,
+            self.markermode,
+            self.bar_positive,
+            self.bar_negative,
+        )
+
+    def is_pct_graph(self):
+        return self.y_label == GraphTitles.PCT
+
+
+class SingleLaneSubplot(Subplot):
+    def __init__(
+            self,
+            title,
+            df,
+            x_col,
+            y_col,
+            y_label,
+            colourby,
+            shapeby,
+            hovertext_cols,
+            cutoff_lines=None,
+            name_col=PINERY_COL.SampleName,
+            markermode="markers",
+            bar_positive=None,
+            bar_negative=None,
+            log_y=False
+    ):
+        super().__init__(
+            title,
+            df,
+            x_col,
+            y_col,
+            y_label,
+            colourby,
+            shapeby,
+            hovertext_cols,
+            cutoff_lines,
+            name_col,
+            markermode,
+            bar_positive,
+            bar_negative,
+            log_y,
+        )
+
+
+class CallReadySubplot(Subplot):
+    def __init__(
+            self,
+            title,
+            df,
+            x_col,
+            y_col,
+            y_label,
+            colourby,
+            shapeby,
+            hovertext_cols,
+            cutoff_lines=None,
+            name_col=PINERY_COL.SampleName,
```

Comment:
Hmm, there seems to be only one place `name_col` is used: 

https://github.com/oicr-gsi/dashi/blob/8b453a9cf24eba57ed659a701f43083396b6e3e4/application/dash_application/utility/plot_builder.py#L347-L356

Should this be replaced by `x=x_fn(data)` to match the other x axis coordinates:
https://github.com/oicr-gsi/dashi/blob/8b453a9cf24eba57ed659a701f43083396b6e3e4/application/dash_application/utility/plot_builder.py#L362
https://github.com/oicr-gsi/dashi/blob/8b453a9cf24eba57ed659a701f43083396b6e3e4/application/dash_application/utility/plot_builder.py#L377
https://github.com/oicr-gsi/dashi/blob/8b453a9cf24eba57ed659a701f43083396b6e3e4/application/dash_application/utility/plot_builder.py#L541

Reply:
name_col originated, I believe, in the early days when there were only Single-Lane graphs, because x was _always_ just PineryColumn.SampleName, so that was an easy way to not have to specify that for the x axis on every graph definition. I'd like for that idea to come back - only specifying the x-axis function if it's not the Single-Lane or Call-Ready default x-axis

Reply:
I will do a PR to remove `name_col` for now then as it would cause a bug if `x_fn` was different from `name_col`. Setting a default for `x_fn` is more annoying, as it would require changing the signature of all views that call `generate` (let's leave that for a broader refactor).

Reply:
See #209